doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+   :language: text
+   :lines: 3
+
+Moreover, the execution will be equivalent to the following
+data-parallel forall-loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 14-15
+
+Nothing about the above requires using a 1D array.  We could also
+promote ``negate()`` by passing it a multidimensional array of reals:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 19-22
+
+which would again be equivalent to the loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 25-26
+
+
+Multi-Argument Promotion
+------------------------
+
+When multiple function arguments are promoted in a single call, the
+result is equivalent to a zippered forall-loop.  As an example,
+consider the following function which takes two ``real`` arguments by
+``ref`` and sorts them relative to one another using a swap
+assignment:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 30-33
+
+This function can be promoted using two arrays of reals as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 35-38
+
+This promotion is equivalent to the following zippered forall-loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 45-46
+
+Either form will leave the arrays as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+   :language: text
+   :lines: 15-16
+
+Promoting multiple arguments simultaneously requires that the
+arguments can legally be zippered together, as in an explicit
+forall-loop.
+
+All of a function's arguments need not be promoted simultaneously.
+Any arguments that are promoted define the implicit, potentially
+zippered forall-loop.  All others are simply passed through to their
+formal argument normally.  For example, consider the following
+function which assigns its *y* argument into its *x* argument only if
+the third argument, *b*, is true:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 51-54
+
+The following four calls demonstrate that various combinations of
+arguments may be promoted, where *Mask* is an array of ``bool``
+values:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 67,76,85,94
+
+These calls are equivalent to the following forall-loops:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 71-72,80-81,89-90,98-99
+
+If *A*, *B*, and *Mask* hold the following values before each of
+these calls:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 63-65
+
+Then *A* will store the following values afterward:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 23,25,27,29
+
```

Comment:
I believe I did this in part to set up for the following section which essentially demonstrates that if either of the latter two arguments are promoted, the first must be as well due to default forall intents.  Does your hesitancy extend even with that explanation?  If so, which variations would you keep and which would you drop?  (I'm about to push a commit that tries to simplify things slightly by moving the definitions of A, B, and Mask before the calls to use three code blocks rather than 4 which I think also helps a little).